llvm/lib/Transforms/IPO/GlobalOpt.cpp

```diff
@@ -2029,6 +2031,109 @@ OptimizeFunctions(Module &M,
   return Changed;
 }
 
+static bool IsCharArray(Type *T) {
+  const unsigned int CHAR_BIT_SIZE = 8;
+  return T && T->isArrayTy() && T->getArrayElementType()->isIntegerTy() &&
+         T->getArrayElementType()->getIntegerBitWidth() == CHAR_BIT_SIZE;
```

Comment:
Yup. I've made the solution generic enough that it can theoretically pad any non-byte array. For ARM it will only attempt padding for arrays of integers. I think this *could* work for more complex types like structs but that might be out of the scope of the PR.